grader_service/handlers/lectures.py

```diff
@@ -106,55 +110,75 @@ async def put(self, lecture_id: int):
         """
-        "Soft"-delete a lecture.
-        Softdeleting: lecture is still saved in the datastore
-        but the users have not access to it.
+        Soft or Hard-Deletes a specific lecture.
+        Soft deleting: lecture is still saved in the datastore
+        but the users have no access to it.
+        Hard deleting: removes lecture from the datastore and all associated directories/files.
 
         :param lecture_id: id of the lecture
         :type lecture_id: int
         :raises HTTPError: throws err if lecture was already deleted
         or was not found
 
         """
-        self.validate_parameters()
+        self.validate_parameters("hard_delete")
+        hard_delete = self.get_argument("hard_delete", "false") == "true"
+
         try:
-            lecture = self.session.get(Lecture, lecture_id)
-            if lecture.deleted == 1:
-                raise HTTPError(404)
-            lecture.deleted = 1
-            a: Assignment
-            for a in lecture.assignments:
-                if (len(a.submissions)) > 0:
-                    self.session.rollback()
-                    raise HTTPError(
-                        HTTPStatus.CONFLICT, "Cannot delete assignment because it has submissions"
-                    )
-                if a.status in ["released", "complete"]:
-                    self.session.rollback()
+            lecture = self.get_lecture(lecture_id)
+            if lecture is None:
+                raise HTTPError(HTTPStatus.NOT_FOUND, reason="Lecture was not found")
+
+            if hard_delete:
+                if not self.user.is_admin:
                     raise HTTPError(
-                        HTTPStatus.CONFLICT,
-                        "Cannot delete assignment because its status is not created",
+                        HTTPStatus.FORBIDDEN, reason="Only Admins can hard-delete lecture."
                     )
 
-                a.deleted = 1
-            self.session.commit()
+                self.delete_lecture_files(lecture)
+                self.session.delete(lecture)
+                self.session.commit()
+            else:
+                if lecture.deleted == DeleteState.deleted:
+                    raise HTTPError(HTTPStatus.NOT_FOUND)
+
+                lecture.deleted = DeleteState.deleted
+                a: Assignment
+                for a in lecture.assignments:
+                    if len(a.submissions) > 0:
+                        self.session.rollback()
+                        raise HTTPError(
+                            HTTPStatus.CONFLICT, "Cannot delete assignments with submissions!"
+                        )
+                    if a.status in ["released", "complete"]:
+                        self.session.rollback()
+                        raise HTTPError(
+                            HTTPStatus.CONFLICT, "Cannot delete released or completed assignments!"
+                        )
+                    a.deleted = DeleteState.deleted
```

Comment:
I believe you also have to handle the case where there already exists a soft-deleted assignment with the same name here. If you try to soft-delete another one, it will violate the unique constraint on `("name", "lectid", "deleted")`.
Or am I missing something?